fix: completing the InternalEncoder for utf8

yosion-p opened the pull request:
Hi bro,It's me, again.🤣
I have completed the `InternalEncoder` for utf8 with surrogates after I read what  you meant(#275 ).
#275 looked a bit messy, so I submitted a brand new one,
I didn't know much before, but now I know more.
I think surrogates should be in pairs always.
Thank you for your patience.

ashtuchkin commented on test/streams-test.js:
```diff
@@ -135,6 +135,12 @@ function checkDecodeStream(opts) {
 describe("Streaming mode", function() {
+  	it("Encoding using internal modules: utf8 with surrogates in separate chunks", checkEncodeStream({
+        encoding: "utf8",
+        input: ["\uD83D", "\uDE3B"],
```
Could you check that strings with more than one character are processed too? E.g. 
```
input: [
    "\uD83D", "\uDE3B", "\n", 
    "abc\uD83D", "\uDE3B", "\n", 
    "\uD83D", "\uDE3Befg", "\n", 
    "abc\uD83D", "\uDE3Befg", "\n", 
],
```

Also, non-matching surrogates at the boundary like low-low, high-high, high-low.

Also, make an input that starts with a low surrogate, but doesn't end with a high surrogate (`input: ["abc\uD83D"]`) - that should probably return the surrogate unchanged.

yosion-p replied:
I've thought of these questions, but not as comprehensively as you.
For example, I thought surrogates should be in pairs always.
In short, I completed some test cases and passed all of them.
It might not be perfect, but it's better.

ashtuchkin reviewed (changes requested):
Looks pretty good! Could you add more tests please (see comment)?

yosion-p commented:
It's confusing that I updated the code but did't run CI by auto.

ashtuchkin commented on encodings/internal.js:
```diff
@@ -204,22 +204,33 @@ InternalDecoderCesu8.prototype.end = function() {
-        this.lowSurrogate = '';
-        return Buffer.from(str, this.enc);
+    var normalStr = '';
+    for (var i = 0; i < str.length; i++) {
```
a loop over the whole string is pretty inefficient. Moreover, you don't need to touch internal surrogates - Node processes them correctly already. We only need to adjust boundaries, as these are what Node can't adjust as it doesn't know about them.

ashtuchkin commented on test/streams-test.js:
```diff
@@ -337,3 +331,46 @@ describe("Streaming sugar", function() {
+    it("surrogates pair be interrupted", checkEncodeStream({
+        encoding: "utf8",
+        input: ["abc\uD83D", "efg\uDE3B"],
+        output: "616263656667",
```
The output we expect should be equivalent to what Node does on a merged input chunks:
```
> new Buffer("abc\uD83D" + "efg\uDE3B", "utf8").toString("hex")
'616263efbfbd656667efbfbd'
```

ashtuchkin reviewed (changes requested):
See comments below.

The behavior we want is to be always equivalent to what Node does, independently of where chunk breaks happen. More formally, for every string S and for every split of S into chunks, the merged output of iconv-lite encoding must be equal to Node's encoding of the whole string (`Buffer.from(S, "utf8")`).

yosion-p commented:
I think I got it,
> the merged output of iconv-lite encoding must be equal to Node's encoding of the whole string (`Buffer.from(S, "utf8")`).

sorry,I thought incomplete surrogate pairs were invalid and could be filtered out.
I will correct all outputs.

But about that:
> a loop over the whole string is pretty inefficient. Moreover, you don't need to touch internal surrogates - Node processes them correctly already. We only need to adjust boundaries, as these are what Node can't adjust as it doesn't know about them.

I don't really understand, 
Because if I don't loop through the string, I won't know which character in the string needs to be processed.

yosion-p commented:
![image](https://user-images.githubusercontent.com/86145334/133398317-a8924c5c-c8de-4ebb-98cb-9cdb7815c19d.png)

I was wondering if there was something wrong with the test method,
Just like I thought before,
If UT into multiple strings in the input,`IconvLiteEncoderStream.prototype._transform` will be invoked many times, back to characters be cut by Encoder rather than a complete surrogates pair.

If we modify InternalEncoder,
When the surrogates pair was cut,
The first call doesn't return anything, only the second one returns a full surrogates pair,

That's why I need to go through a loop and check for characters one by one to see if it is a surrogate.

yosion-p commented:
I think I can solve this problem  in `InternalEncoderUtf8.prototype.end()`,
I will try.

yosion-p commented:
Hi bro,It's me, again.🤣
I did it by modifying the code of `InternalEncoderUtf8.prototype.write`.
In a word, less is more.

ashtuchkin commented on test/streams-test.js:
```diff
@@ -331,3 +331,52 @@ describe("Streaming sugar", function() {
+        input: ["\uDE3Befg"],
+        output: Buffer.from("\uDE3Befg", "utf8").toString("hex")
+    }));
+});
```
could you also check the following input, I think current algorithm might be wrong on it:
```
["\uD83D", "\uDE3B\uD83D", "\uDE3B"]
```

ashtuchkin reviewed (commented):
Getting close, left a couple more comments.

yosion-p commented on encodings/internal.js:
```diff
@@ -196,3 +198,30 @@ InternalDecoderCesu8.prototype.end = function() {
+        this.surrogateLeftStr = str
+        // do nothing
+    } else {
+        var buf = Buffer.from(this.surrogateLeftStr + str)
```
uhh,i saw that,sorry,it's my oversight.

yosion-p commented:
have a look😃

ashtuchkin commented on encodings/internal.js:
```diff
@@ -196,3 +198,30 @@ InternalDecoderCesu8.prototype.end = function() {
+InternalEncoderUtf8.prototype.write = function (str) {
+    var lastr = str[str.length - 1]
+    if ('\uD800' < lastr && lastr <= '\uDBFF') {
+        this.surrogateLeftStr += str
```
this has the potential of accumulating unlimited data in this.surrogateLeftStr as we stream.. Can we encode and return everything but the last character instead?

ashtuchkin commented on test/streams-test.js:
```diff
@@ -331,3 +326,71 @@ describe("Streaming sugar", function() {
+    it("a single string", checkEncodeStream({
+        encoding: "utf8",
+        input: ["\uD83D\uDE3B"],
+        output: toBufForSurrgates(["\uD83D\uDE3B"])
```
you still duplicate the input values here, can we avoid that and use 1 line per test instead of 5?

ashtuchkin commented on test/streams-test.js:
```diff
@@ -331,3 +326,71 @@ describe("Streaming sugar", function() {
 
+describe("Encoding using internal modules with surrogates in separate chunks:", function () {
+    // Generate the expected value of surrogates UT
+    function toBufForSurrgates(input) {
```
nit: typo in work "surrogates"

yosion-p commented:
Aha i tried to make it short, but didn't work it out.
> nit: typo in work "surrogates"

ashtuchkin commented on encodings/internal.js:
```diff
@@ -196,3 +198,34 @@ InternalDecoderCesu8.prototype.end = function() {
+        var buf = Buffer.from(this.lowSurrogate + str, this.enc);
+        this.lowSurrogate = ''
+        return buf
+    }
```
I *think* this should work correctly, but can we streamline it even more?
 1. Can you make this function have only one `Buffer.from(` and one string concatenation `this.lowSurrogate + str`? The branches of this if statement have some repetition in them and I think removing it would make the code clearer.
 2. `'\uD800' < lastr && lastr <= '\uDBFF'` is a string comparison and is afaik expensive. Can we maybe make `lastr` and `this.lowSurrogate` integers using `s.charCodeAt` and switch to an integer comparison? Also could you name `lastr` better? I'm not sure what the current name means.
 3. Please make sure the function does not throw if str is empty.

ashtuchkin commented on test/streams-test.js:
```diff
@@ -331,3 +326,33 @@ describe("Streaming sugar", function() {
 
+describe("Encoding using internal modules with surrogates in separate chunks:", function () {
+    // Generate the expected object of surrogates UT,by input array
+    function generatesrgObj(input) {
```
I don't think you can use `generatesrgObj` in any way except by passing its result to `checkEncodeStream`, right? In that case why not move `checkEncodeStream` inside like this?
```
function checkUtf8EncodeStream(input) {
  return checkEncodeStream({
    encoding: "utf8",
    input: input,
    ...
```

ashtuchkin reviewed (commented):
Almost there, thank you for persistence!

yosion-p commented:
> Almost there, thank you for persistence!

Victory lies ahead!

ashtuchkin commented on test/streams-test.js:
```diff
@@ -331,3 +329,54 @@ describe("Streaming sugar", function() {
+    it("a single string", checkEncodeStream({
+        encoding: "utf8",
+        input: ["\uD83D\uDE3B"]
+    }))
```
could you explain why you've decided to get back to calling checkEncodeStream directly? Did my previous suggestion not work?

ashtuchkin commented on encodings/internal.js:
```diff
@@ -196,3 +198,34 @@ InternalDecoderCesu8.prototype.end = function() {
+        str = this.lowSurrogate + str
+        this.lowSurrogate = ''
+    }
+    return Buffer.from(str, this.enc);
```
I see you're returning Buffer.from in the end, this is good.
The `lastr` handling became more messy though. I thought we could do something like this:
```
InternalEncoderUtf8.prototype.write = function (str) {
    if (!str) return;

    if (this.lowSurrogate) {
        str = this.lowSurrogate + str;
        this.lowSurrogate = '';
    }

    var lastCharCode = str.charCodeAt(str.length-1);
    if (0xD800 < lastCharCode && lastCharCode <= 0x...) {
        this.lowSurrogate = str.slice(str.length - 1);
        str = str.slice(0, str.length - 1)
    }

    return Buffer.from(str, this.enc);
}
```

yosion-p commented:
Hi Buddy, 
> I see all the tests have the same structure ....

Now i realize actually you made it very clear, it's me misunderstand it, I'm terribly sorry…

> make sure the function does not throw if str is empty.

About what you said,
I have my own interpretation,
I added one  condition judgment was also in the first line of `InternalEncoderUtf8.prototype.write` , then I removed.

I think we still need to consider the incoming string is empty, as `internal. js line: 81 InternalEncoder.prototype.write`,
otherwise UT will report an error, like I just submitted for the first time.

ashtuchkin commented:
I agree, we need to return an empty buffer, not undefined, when str is empty.

I've made some final minor adjustments (e.g. it's actually a high surrogate, not low :) and will merge. Thank you for your contribution!

yosion-p commented:
Thank you buddy! It's a milestone for me the first time my code got merged in an open source project. 




Even though my pointcut and idea are right since the beginning, but the process was kind of tortuous. 




Maybe I need to pay more attention to the format and performance of the code,

and try to get your full point.lol




In short, thanks a lot for your patience and encouragement, respect!



Best Regards,
Yosion.


At 2021-09-19 01:55:57, "Alexander Shtuchkin" ***@***.***> wrote:

I agree, we need to return an empty buffer, not undefined, when str is empty.

I've made some final minor adjustments (e.g. it's actually a high surrogate, not low :) and will merge. Thank you for your contribution!

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

ashtuchkin commented:
You did great, keep it up! It all comes with practice, don't worry too much.
